### DISCUSSION TITLE
Как отправить Email с прикрепленной таблицей

### DISCUSSION THREAD

**Discussion by vldekov** [totumonline/totum-ru-issues-and-discussions#544] [Как это сделать?] [answered]
Есть задача отправить Email  с прикрепленной таблицей.
1. Как сделать чтобы в тело письма записалась таблица?
Например таблица записывается в поле h_data1 формата Данные.
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/114401808/0be8525f-4b3c-4f64-8f4c-baa230180a81)
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/114401808/c3d8fbe7-dc14-4b67-962f-45e70d85ae68)

2. Как сделать чтобы к письму прикрепился файл ( например, как результат выгрузки таблицы в CSV файл )?

**Comment by Re-Requeste** [comment, marked as answer]
Здравствуйте. 

1. Собрал бы строчные части в отдельной ячейке использованием синтакисов html
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/137899742/ecb465c4-c616-4c98-800e-d8176a15a419)

2. Через selectList собрал бы нужные мне строки с использованием фильтров и отправил бы все через emailSend
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/137899742/d7ef5d9d-1ecd-41cd-8761-e86b6b31e2bf)

3. На почту придет сообщение с телом таблицы, естественно все линий и т.п. сами настроите в html + css 
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/137899742/bd94ad80-9bc1-4364-9980-8c6909f823d6)


По поводу файла, думаю таким же образом можно создать файл .txt и внутрь закинуть строки, разделив столбцы между собой,
Вот пример из документации: 

emailSend(to: "test@example.ru"; title: "Отправка письма"; body: "Тестовое письмо"; files: json`[$filerow]`)
    fileRow: rowCreate(field: "name" = "Текст файл.txt"; field: "filestring" = $text)
        text: "Какой-то текст"

**Reply by vldekov** [reply]
Огромное человеческое СПАСИБО !!!

**Reply by erpic-ru** [reply]
Такое почти всегда через шаблоны печати делаю
```
= : emailSend(to: $email; title: "Подходит срок выполнения ТО"; body: $body; )

body: textByTemplate(template: "to_main"; data: $data)

data: rowCreate(field: "cars" = $innerrowlist;)
// Для повторяющихся позиций используем вложенный шаблон "to", передавая в него строчную часть таблицы.

innerrowlist: rowCreate(field: "template" = "to"; field: "data" = $rl)
rl: selectRowList(table: 'tyagachi'; field: 'nomer_ts'; sfield: 'servis';  field: 'win'; field: 'probeg';  sfield: 'vid_rabot';  field: 'probeg_next';field: 'diff' ; where: 'sobstvennost' = $s; where: 'probeg' != ""; where: 'diff' <= 5000; order: 'nomer_ts' asc)
```

Ну и в шаблонах уже html и css

**Reply by Re-Requeste** [reply]
Это более правильный вариант я б сказал

**Reply by vldekov** [reply]
А можно попросить прислать шаблоны "to_main" и "to" ?

**Reply by erpic-ru** [reply]
Там какой-то кривейший html, но для понимания пойдет:)

**to_main:**

```
<table>
  <thead>
	<tr style="border: solid 1px #474747; padding: 5px;">
	  <th style="border: solid 1px #474747; padding: 5px;">No
	  </th>
	  <th style="border: solid 1px #474747; padding: 5px;">VIN
	  </th>
	  <th style="border: solid 1px #474747; padding: 5px;">Текущий пробег
	  </th>
	  <th style="border: solid 1px #474747; padding: 5px;">Пробег для замены
	  </th>
	  <th style="border: solid 1px #474747; padding: 5px;">Осталось до замены
	  </th>
	  <th style="border: solid 1px #474747; padding: 5px;">Вид работ
	  </th>
	  <th style="border: solid 1px #474747; padding: 5px;">Сервис
	  </th>
	</tr>
  </thead>
  <tbody>
	{cars}	
  </tbody>
  </table>
```

**to**:
```
<tr>
  <td style="border: solid 1px #474747; padding: 5px;">{nomer_ts}
  </td>
<td style="border: solid 1px #474747; padding: 5px;">{win}
  </td>
<td style="border: solid 1px #474747; padding: 5px;">{probeg} км.
  </td>
<td style="border: solid 1px #474747; padding: 5px;">{probeg_next} км.
  </td>
<td style="border: solid 1px #474747; padding: 5px;">{diff} км.
  </td>
<td style="border: solid 1px #474747; padding: 5px;">{vid_rabot}
  </td>
  <td style="border: solid 1px #474747; padding: 5px;">{servis}
  </td>
</tr>
```

**Reply by ttm-alexeypolunin** [reply]
Давно уже в списке на разработку функция strRowListToHtml, тк это одна из самых распространенных задач (отправить табличку в теле письма) и мне например дико геморно это делать через шаблоны.

**Comment by vldekov** [comment]
Всем спасибо! Очень помогли.
